src/twisted/web/test/test_agent.py

```
@@ -3216,6 +3216,24 @@ def test_cancel(self):
         self.failureResultOf(deferred, defer.CancelledError)
         self.assertTrue(response.transport.aborting)
 
+    def test_cancelWithConnectionLost(self):
+        """
+        When cancelling the L{Deferred} returned by L{client.readBody}, the
+        CancelledError failure is sent repeatedly in response protocol. Exception
+        defer.AlreadyCalledError should be suppressed
+        """
+        response = DummyResponse()
+        deferred = client.readBody(response)
+        deferred.cancel()
+
+        try:
+            response.protocol.connectionLost(Failure(defer.CancelledError))
+        except defer.AlreadyCalledError:
```

Why do we need to catch the exception as part of the test execution?

---

I didn't find better way to mark the test case as [FAILED] in case if the exception is thrown. If the exception is not handled the test case will be marked as [ERROR]. Not much difference but it seems more properly

---

Remove the catch?

---

I think `[ERROR]` is likely to be better-behaved in terms of traceback-reporting behavior, since `AlreadyCalledError` could always be raised out of a *different* code path than the one we were expecting, which would not exactly what we would try to be asserting about. I can see why you'd want to get a `[FAILED]` given that this is the existing behavior, but in this case it's better to just let the exception propagate so that it's easier to debug.